### PR TITLE
[GeoMechanicsApplication] Extract plane strain and three dimensional stress state functionality

### LINKED ISSUE #12004: [GeoMechanicsApplication] Extract element functionality related to stress state using a strategy pattern
As a developer, I want the stress state functionality to be separated from the elements, such that the elements become more maintainable and extendible and easier to understand.

#### Acceptance criteria
- All functionality related to stress states, which could have different implementations (i.e. axisymmetrical, plane strain and full-3d) is moved to a separate class using the strategy design pattern see e.g. [wikipedia](https://en.wikipedia.org/wiki/Strategy_pattern) for a bit more info. More concretely, this includes the following functions `CalculateBMatrix`, `CalculateIntegrationCoefficient`, `CalculateCauchyGreenStrain` and `CalculateCauchyAlmansiStrain`.
- Three strategies/policies are created, all deriving from an abstract interface, to provide the stress-state functionality.

For some experimentation done on the topic, see the following PR: https://github.com/KratosMultiphysics/Kratos/pull/11966

### PR DESCRIPTION
**Pull request by rfaasse**
**📝 Description**
This PR aims to extract the plane strain and three dimensional stress state functionality, such that the element does not know about the implementation for the different flavours.

This includes:
- Creating the ThreeDimensionalStressState and PlaneStrainStressState with implementations for `CalculateBMatrix`, `CalculateIntegrationCoefficient` and `CalculateGreenLagrangeStrain` and a `Clone` function, needed due to Kratos' registration system.
- Add unit tests to test if the implementations are correct
- Use the calculation functions mentioned above in the elements.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp [hunk 1]
```diff
@@ -700,7 +702,15 @@ double UPwBaseElement<TDim, TNumNodes>::CalculateIntegrationCoefficient(
...
-    return IntegrationPoints[PointNumber].Weight() * detJ;
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if constexpr (TDim == 2) {
+        stress_state_policy = std::make_unique<PlaneStrainStressState>();
+    } else {
+        stress_state_policy = std::make_unique<ThreeDimensionalStressState>();
+    }
+
+    return stress_state_policy->CalculateIntegrationCoefficient(IntegrationPoints[PointNumber],
+                                                                detJ, GetGeometry());
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: I know that the calculation of the integration coefficient is the same for plane strain and 3D, so this might seem overkill. However, this makes it consistent with all the other stress-state dependent functions, such that making the stress state policy a member of the elements becomes simpler.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 2]
```diff
@@ -1209,33 +1211,15 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateBMatrix(Matrix& rB, const
...
-            rB(INDEX_3D_ZZ, index + INDEX_Z) = GradNpT(i, INDEX_Z);
-            rB(INDEX_3D_XY, index + INDEX_X) = GradNpT(i, INDEX_Y);
-            rB(INDEX_3D_XY, index + INDEX_Y) = GradNpT(i, INDEX_X);
-            rB(INDEX_3D_YZ, index + INDEX_Y) = GradNpT(i, INDEX_Z);
-            rB(INDEX_3D_YZ, index + INDEX_Z) = GradNpT(i, INDEX_Y);
-            rB(INDEX_3D_XZ, index + INDEX_X) = GradNpT(i, INDEX_Z);
-            rB(INDEX_3D_XZ, index + INDEX_Z) = GradNpT(i, INDEX_X);
-        }
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if constexpr (TDim == 2) {
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:1215]: This if-statement is there a couple of times for the elements now. This will all disappear when we pass the policy as a member (then we just delegate the call and the type of policy will have been decided upon construction).

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+std::unique_ptr<StressStatePolicy> ThreeDimensionalStressState::Clone() const
+{
+    return std::make_unique<ThreeDimensionalStressState>();
+}
+
+Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: @indigocoral I added a utility function to calculate the ETensor, which we now re-use for 3D and plane strain.

File: applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp [hunk 4]
```diff
@@ -132,6 +132,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStrainUtilities
         KRATOS_CATCH("")
     }
 
+    static Matrix CalculateGreenLagrangeStrainTensor(const Matrix& rDeformationGradient);
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp:135]: Here also all other functions can be split, such that we do not need a .hpp but a cleaner .h and .cpp files. I'd propose to do that in a separate PR.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         GradNpT,
+                                                const Vector&         Np,
+                                                const Geometry<Node>& rGeometry) const
+{
+    const auto dimension       = rGeometry.WorkingSpaceDimension();
+    const auto number_of_nodes = rGeometry.size();
+    Matrix     result = ZeroMatrix(VOIGT_SIZE_2D_AXISYMMETRIC, dimension * number_of_nodes);
+
+    for (unsigned int i = 0; i < number_of_nodes; ++i) {
+        const auto index = dimension * i;
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: this is more an offset than an index

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+namespace Kratos
+{
+
+Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         GradNpT,
+                                                const Vector&         Np,
+                                                const Geometry<Node>& rGeometry) const
+{
+    const auto dimension       = rGeometry.WorkingSpaceDimension();
+    const auto number_of_nodes = rGeometry.size();
+    Matrix     result = ZeroMatrix(VOIGT_SIZE_2D_AXISYMMETRIC, dimension * number_of_nodes);
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:22]: I expected alignment of the =, but that may have been overruled by the length of the line.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+Matrix ThreeDimensionalStressState::CalculateBMatrix(const Matrix&         GradNpT,
+                                                     const Vector&         Np,
+                                                     const Geometry<Node>& rGeometry) const
+{
+    const auto dimension       = rGeometry.WorkingSpaceDimension();
+    const auto number_of_nodes = rGeometry.size();
+    Matrix     result          = ZeroMatrix(VOIGT_SIZE_3D, dimension * number_of_nodes);
+
+    for (unsigned int i = 0; i < number_of_nodes; ++i) {
+        auto index = dimension * i;
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: Also this is more an offset than an index.

File: applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp (hunk 4 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp:135]: Yes please

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp [hunk 8]
```diff
@@ -17,34 +17,21 @@
...
-}
-
 KRATOS_TEST_CASE_IN_SUITE(CalculateBMatrixWithValidGeometryReturnsCorrectResults, KratosGeoMechanicsFastSuite)
 {
     const std::unique_ptr<StressStatePolicy> p_stress_state_policy =
         std::make_unique<AxisymmetricStressState>();
 
     Model model;
-    auto& model_part = CreateModelPart(model);
+    auto& model_part = ModelSetupUtilities::CreateModelPartWithASingle2D3NElement(model);
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp]: Should the variable name start with r?

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Kratos
+{
+
+class ThreeDimensionalStressState : public StressStatePolicy
+{
+public:
+    [[nodiscard]] Matrix CalculateBMatrix(const Matrix&         GradNpT,
+                                          const Vector&         Np,
+                                          const Geometry<Node>& rGeometry) const override;
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.h]: Does this follow the naming conventions?

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+//
+#include "three_dimensional_stress_state.h"
+#include "custom_utilities/stress_strain_utilities.hpp"
+
+namespace Kratos
+{
+
+Matrix ThreeDimensionalStressState::CalculateBMatrix(const Matrix&         GradNpT,
+                                                     const Vector&         Np,
+                                                     const Geometry<Node>& rGeometry) const
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: Does this follow the naming conventions?

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 11]
```diff
@@ -2278,32 +2264,14 @@ Vector SmallStrainUPwDiffOrderElement::CalculateGreenLagrangeStrain(const Matrix
...
-    if (Dim == 2) {
-        Vector StrainVector;
-        StrainVector = MathUtils<double>::StrainTensorToVector(ETensor);
-        result[INDEX_2D_PLANE_STRAIN_XX] = StrainVector[0];
-        result[INDEX_2D_PLANE_STRAIN_YY] = StrainVector[1];
-        result[INDEX_2D_PLANE_STRAIN_ZZ] = 0.0;
-        result[INDEX_2D_PLANE_STRAIN_XY] = StrainVector[2];
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if (GetGeometry().WorkingSpaceDimension() == 2) {
+        stress_state_policy = std::make_unique<PlaneStrainStressState>();
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp]: Should these start with a p?

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
+
+    const auto StrainVector          = MathUtils<double>::StrainTensorToVector(ETensor);
+    Vector     result                = ZeroVector(VOIGT_SIZE_2D_PLANE_STRAIN);
+    result[INDEX_2D_PLANE_STRAIN_XX] = StrainVector[0];
+    result[INDEX_2D_PLANE_STRAIN_YY] = StrainVector[1];
+    result[INDEX_2D_PLANE_STRAIN_ZZ] = 0.0;
+    result[INDEX_2D_PLANE_STRAIN_XY] = StrainVector[2];
+
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: I think this part of line 50 to 54 I've seen repeatedly. Should it therefore become something in the stress_strain utitlities.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp [hunk 13]
```diff
@@ -700,7 +702,15 @@ double UPwBaseElement<TDim, TNumNodes>::CalculateIntegrationCoefficient(
     const GeometryType::IntegrationPointsArrayType& IntegrationPoints, unsigned int PointNumber, double detJ)
 
 {
-    return IntegrationPoints[PointNumber].Weight() * detJ;
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if constexpr (TDim == 2) {
+        stress_state_policy = std::make_unique<PlaneStrainStressState>();
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: p_stress_state_policy ?

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+#include "stress_state_policy.h"
+
+namespace Kratos
+{
+
+class PlaneStrainStressState : public StressStatePolicy
+{
+public:
+    [[nodiscard]] Matrix CalculateBMatrix(const Matrix&         GradNpT,
+                                          const Vector&         Np,
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h]: rGradNpT and rNp ?

**Review by WPK4FEM — COMMENTED** [review]
I looks like this will get rid of a lot of duplicated code and provides tested basic functionality.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 6 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:22]: Indeed, aligning it would result in a width of 102 characters unfortunately

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 5 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Good point, changed the naming!

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 12 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: I agree, but in my opinion the `PlaneStrainStressState` is the only place it should be (converting a strain tensor to a vector). For example, the Hencky strain calculation could also move here in the same way as for Green Lagrange: The calculation of the tensor would be done in StressStrainUtilities and the conversion to the plain strain vector would be done in this class. What is your opinion on this?

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h (hunk 14 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h]: Good catch! Changed

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp (hunk 11 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp]: Agreed, changed all of them to p_stress_state_policy

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp (hunk 8 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp]: Walked through the axisymmetric and plane strain tests and changed the model_part that I could find to r_model_part. Could you do the same for the 3D tests @indigocoral?

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp (hunk 13 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: Yes, done

**Review by rfaasse — COMMENTED** [review]
Thanks for the quick review @WPK4FEM! I processed the comments except for the ones related to the 3D stress state, which @indigocoral will pick up.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp (hunk 2 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:1215]: Thanks for explaining. Agreed.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 12 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Indeed, the tensor based part in a utility, then the vector part that is specific for the stress state in the stress state functionality.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        result(INDEX_2D_PLANE_STRAIN_YY, offset + INDEX_Y) = rGradNpT(i, INDEX_Y);
+        result(INDEX_2D_PLANE_STRAIN_XY, offset + INDEX_X) = rGradNpT(i, INDEX_Y);
+        result(INDEX_2D_PLANE_STRAIN_XY, offset + INDEX_Y) = rGradNpT(i, INDEX_X);
+    }
+
+    return result;
+}
+
+double PlaneStrainStressState::CalculateIntegrationCoefficient(const Geometry<Node>::IntegrationPointType& rIntegrationPoint,
+                                                               double detJ,
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Nitpicking (to follow the Kratos Style Guide):
```suggestion
                                                               double DetJ,
```

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+                                                               const Geometry<Node>&) const
+{
+    return rIntegrationPoint.Weight() * detJ;
+}
+
+Vector PlaneStrainStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Perhaps naming the result `strain_tensor` is clearer than `ETensor`? (And it would be more in line with the variable name in the next statement: `StrainVector`, which should be `strain_vector`.) In fact, since the tensor form is not used (we only use the vector form), we may consider to merge the two statements, i.e.:
```
const auto strain_vector = MathUtils<>::StrainTensorToVector(StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient));
```

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+class PlaneStrainStressState : public StressStatePolicy
+{
+public:
+    [[nodiscard]] Matrix CalculateBMatrix(const Matrix&         rGradNpT,
+                                          const Vector&         rNp,
+                                          const Geometry<Node>& rGeometry) const override;
+    [[nodiscard]] double CalculateIntegrationCoefficient(const Geometry<Node>::IntegrationPointType& rIntegrationPoint,
+                                                         double detJ,
+                                                         const Geometry<Node>& rGeometry) const override;
+    [[nodiscard]] Vector CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const override;
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h]: After a brief discussion with Wijtze Pieter, we feel that a better parameter name would be `rDeformationGradient` (i.e. stripping of the `Total` part).

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp (hunk 10 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: In addition to Wijtze Pieter's comment, I would suggest to remove the name `Np` here, since it is not being used at all in this override. Please, also remove the parameter name in the corresponding header file.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+//  Main authors:    Richard Faasse
+//
+#include "plane_strain_stress_state.h"
+#include "custom_utilities/stress_strain_utilities.hpp"
+
+namespace Kratos
+{
+
+Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         rGradNpT,
+                                                const Vector&         rNp,
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: I would suggest to make this parameter unnamed (i.e. remove `rNp`), since it is not being used by this override. Please, also propagate that change to the header file.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        result(INDEX_3D_YZ, index + INDEX_Z) = GradNpT(i, INDEX_Y);
+        result(INDEX_3D_XZ, index + INDEX_X) = GradNpT(i, INDEX_Z);
+        result(INDEX_3D_XZ, index + INDEX_Z) = GradNpT(i, INDEX_X);
+    }
+
+    return result;
+}
+
+double ThreeDimensionalStressState::CalculateIntegrationCoefficient(
+    const Geometry<Node>::IntegrationPointType& rIntegrationPoint, double detJ, const Geometry<Node>&) const
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: More nitpicking:
```suggestion
    const Geometry<Node>::IntegrationPointType& rIntegrationPoint, double DetJ, const Geometry<Node>&) const
```

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+{
+    return rIntegrationPoint.Weight() * detJ;
+}
+
+Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: As indicated before, please rename this function parameter:
```suggestion
Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rDeformationGradient) const
```

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: See my previous remarks about the name `ETensor` and the possibility to merge this statement with the next one.

File: applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp (hunk 4 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp:135]: Agreed.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_plane_strain_stress_state.cpp [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+#include <boost/numeric/ublas/assignment.hpp>
+
+using namespace Kratos;
+
+namespace Kratos::Testing
+{
+
+KRATOS_TEST_CASE_IN_SUITE(CalculateBMatrixGivesCorrectResults, KratosGeoMechanicsFastSuite)
+{
+    auto stress_state = std::make_unique<PlaneStrainStressState>();
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_plane_strain_stress_state.cpp]: Nitpicking:
```suggestion
    const auto p_stress_state = std::make_unique<PlaneStrainStressState>();
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+    KRATOS_CHECK_MATRIX_NEAR(calculated_matrix, expected_matrix, 1e-12)
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ThreeDimensionalStressState_ReturnsCorrectIntegrationCoefficient, KratosGeoMechanicsFastSuite)
+{
+    const auto p_stress_state_policy = std::make_unique<ThreeDimensionalStressState>();
+
+    Model model;
+    auto& model_part = ModelSetupUtilities::CreateModelPartWithASingle2D3NElement(model);
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp]: It may be confusing to generate a 2D element for this test, since it aims at a 3D stress state.

**Review by avdg81 — COMMENTED** [review]
This is a very clear and clean PR. The hard work that you (@rfaasse and @indigocoral) have put into this is much appreciated, and I'm sure it will pay off in the near future. A big thank you! I only have several minor comments, so I believe we can merge this to `master` soon.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 16 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: The symbol E is used for the GreenLagrange strain tensor in literature. So I disagree with Anne here.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 18 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Good point, done

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 15 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Done

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 16 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: I extracted a function that converts the strain tensor to a vector. Since that one can be used later for more than only Green Lagrange (e.g. for Hencky Strain), I made the argument name there `rStrainTensor`. Is this fine for everyone?

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h (hunk 17 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h]: Done!

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_plane_strain_stress_state.cpp (hunk 22 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_plane_strain_stress_state.cpp]: Good catch!

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+#include <boost/numeric/ublas/assignment.hpp>
+
+using namespace Kratos;
+
+namespace Kratos::Testing
+{
+
+KRATOS_TEST_CASE_IN_SUITE(ThreeDimensionalStressState_CalculateBMatrixReturnsCorrectResults, KratosGeoMechanicsFastSuite)
+{
+    auto stress_state = std::make_unique<ThreeDimensionalStressState>();
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp]: Similar to @avdg81's comment for the plane strain test, this one can also be p_stress_state

**Review by rfaasse — COMMENTED** [review]
Processed the comments for the plane/axi states/tests. @indigocoral and I will have a look tomorrow at the comments on 3D.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp [hunk 25]
```diff
@@ -36,29 +34,31 @@ Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         rGradNpT,
...
-    return result;
+    return ConvertStrainTensorToVector(StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rDeformationGradient));
 }
 
 std::unique_ptr<StressStatePolicy> PlaneStrainStressState::Clone() const
 {
     return std::make_unique<PlaneStrainStressState>();
 }
 
+Vector PlaneStrainStressState::ConvertStrainTensorToVector(const Matrix& rStrainTensor)
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:53]: Nice!

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 16 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Yes, I'm fine with that parameter name.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp (hunk 8 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp]: On behalf of @indigocoral it has been done.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.h (hunk 9 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.h]: Good catch. The parameter names have been updated.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp (hunk 10 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: Both remarks have been addressed.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp (hunk 7 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: The variable name has been changed. Now it is also consistent with the plane strain implementation.

File: applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp (hunk 16 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp]: Ok

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp (hunk 23 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp]: Element was changed to a 3D element

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp (hunk 24 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp]: Has become p_stress_state_policy.

File: applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp (hunk 19 above)
Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp]: Was done

**Review by WPK4FEM — APPROVED** [review]
Tried to go over the comments 1 by 1 and looked in the last version of the code. I'm happy with the current state.
